codejail_service/startup_check.py

```diff
@@ -0,0 +1,124 @@
+"""
+State and accessors for a safety check that is run at startup.
+"""
+
+import logging
+
+from codejail_service.codejail import safe_exec
+
+log = logging.getLogger(__name__)
+
+# Results of the safety check that was performed at startup.
+#
+# Expected values:
+#
+# - None: The check has not yet been performed
+# - True: No issues were detected
+# - False: A fundamental safety or function issue was detected
+#
+# Any other value indicates that the safety check failed in an
+# unexpected way, perhaps raising an exception.
+#
+# The *only value* that indicates that it is safe to receive code-exec
+# calls is `True`.
+STARTUP_SAFETY_CHECK_OK = None
+
+
+def is_exec_safe():
+    """
+    Return True if and only if it is safe to accept code-exec calls.
+    """
+    return STARTUP_SAFETY_CHECK_OK is True
+
+
+def run_startup_safety_check():
+    """
+    Perform a sandboxing safety check.
+
+    Determines if the service is running with an acceptable configuration.
+    This is *not* a full test suite, just a basic check that codejail
+    is actually configured in sandboxing mode.
+
+    This just initializes state. Afterwards, is_exec_safe can be called.
+    """
+    global STARTUP_SAFETY_CHECK_OK
+
+    # App initialization can happen multiple times; just run checks once.
+    if STARTUP_SAFETY_CHECK_OK is not None:
+        return
+
+    tests = [
+        {
+            "name": "Basic code execution",
+            "fn": _test_basic_function,
+        },
+        {
+            "name": "Block sandbox escape by disk access",
+            "fn": _test_escape_disk,
+        },
+        {
+            "name": "Block sandbox escape by child process",
+            "fn": _test_escape_subprocess,
+        },
+    ]
+
+    any_failed = False
+    for test in tests:
+        try:
+            result = test['fn']()
+        except BaseException as e:
+            result = f"Uncaught exception from test: {e!r}"
+
+        if result is True:
+            log.info(f"Startup test {test['name']!r} passed")
+        else:
+            any_failed = True
+            log.error(f"Startup test {test['name']!r} failed with: {result!r}")
+
+    STARTUP_SAFETY_CHECK_OK = not any_failed
+
+
+def _test_basic_function():
```

Comment:
Is there any non-mocked test for this function, and is there any way to? Same for the other calls.

Reply:
Yes -- `test_logging` and `test_unsafe_tests_default` both expect these to actually run.

Reply:
I see. Although maybe this should have been obvious, maybe in `test_logging` you could explain that because we don't actually have any protections enabled via AppArmor in the unit test, we can assume that all safety tests should fail. 

Can you confirm that we don't have any tests that pass the safety tests by mocking at the `safe_exec` level? If not, would this be useful?

Reply:
Sure, I can add that note to the test.

Having trouble understanding the second para.

Reply:
For second paragraph, how to you feel about renaming `test_failure_modes` to `test_success_and_failure_modes` and calling it a day?

Reply:
Ah, I see what you meant now. Yeah, it was just folded in as a baseline, but I should make it more explicit. I'll do the renaming you suggest, but also add more comments and docstrings to make it more clear what `responses` is supposed to be, etc.